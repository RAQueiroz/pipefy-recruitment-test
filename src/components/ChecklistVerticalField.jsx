import React from "react";
import styled from "styled-components";
import StyledLabel from "./StyledLabel";
import { InlineError } from "./StyledElements";

const StyledCheckbox = styled.div`
  border: solid 1px ${({ checked }) => (checked ? "#006699" : "#CCC")};
  width: 20px;
  height: 20px;
  border-radius: 3px;
  margin-right: 8px;
  display: inline-block;
  &:before {
    content: "âœ”";
    position: relative;
    top: 0px;
    left: 3px;
    color: ${({ checked }) => (checked ? "#006699" : "#FFF")};
  }
`;

class ChecklistVerticalField extends React.Component {
  onChange = item => {
    const { value = [], onChange, id } = this.props;

    let index = value.indexOf(item);
    let selected =
      index === -1 ? [...value, item] : value.filter(x => x !== item);

    onChange({ id, value: selected });
  };
  render() {
    const { label, id, error, options, value = [] } = this.props;
    return (
      <div id={id}>
        <StyledLabel>{label}</StyledLabel>
        {options.map((item, index) => (
          <label
            data-testid={`check-${index}`}
            key={item}
            style={{ display: "block", padding: 10 }}
            onClick={event => this.onChange(item)}
          >
            <StyledCheckbox checked={value.indexOf(item) !== -1}  />
            {item}
          </label>
        ))}
        {error && <InlineError>{error}</InlineError>}
      </div>
    );
  }
}

export default ChecklistVerticalField;
