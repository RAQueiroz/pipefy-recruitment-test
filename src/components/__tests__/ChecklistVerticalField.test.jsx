import React from "react";
import ChecklistVerticalField from "../ChecklistVerticalField";
import { render, Simulate, wait } from "react-testing-library";
// import "dom-testing-library/extend-expect";

describe(" <ChecklistVerticalField/>", () => {
  const item = {
    id: "xpto",
    label: "Name",
    value: "",
    options: ["A", "B", "C"]
  };

  const onChangeSpy = jest.fn();
  const { getByText, getByTestId, container, queryByText } = render(
    <ChecklistVerticalField onChange={onChangeSpy} {...item} />
  );

  it("renders", () => {
    const submitButton = queryByText("Name");
    expect(submitButton).toBeTruthy(); // it exist
  });

  it("render correct number of boxes", () => {
    item.options.forEach((item, index) => {
      expect(getByTestId(`check-${index}`)).toBeTruthy();
    });
  });

  it("match snapshot", () => {
    expect(container.firstChild).toMatchSnapshot();
  });

  it("on change to be called with the correct values", () => {
    Simulate.click(getByTestId(`check-1`));
    expect(onChangeSpy).toHaveBeenCalledWith({ id: "xpto", value: ["B"] });
  });
});
