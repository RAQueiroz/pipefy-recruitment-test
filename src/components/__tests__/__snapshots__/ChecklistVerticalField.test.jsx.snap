// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` <ChecklistVerticalField/> match snapshot 1`] = `
<div
  id="xpto"
>
  <label
    class="sc-bdVaJa hOggSJ"
  >
    Name
  </label>
  <label
    data-testid="check-0"
    style="display: block; padding: 10px;"
  >
    <div
      class="sc-gzVnrw bIfHpe"
    />
    A
  </label>
  <label
    data-testid="check-1"
    style="display: block; padding: 10px;"
  >
    <div
      class="sc-gzVnrw bIfHpe"
    />
    B
  </label>
  <label
    data-testid="check-2"
    style="display: block; padding: 10px;"
  >
    <div
      class="sc-gzVnrw bIfHpe"
    />
    C
  </label>
</div>
`;
